packages/api/src/external/hie/carequality-analytics.ts

```diff
@@ -0,0 +1,191 @@
+import {
+  OutboundPatientDiscoveryResp,
+  OutboundDocumentQueryResp,
+  OutboundDocumentRetrievalResp,
+  isSuccessfulOutboundDocQueryResponse,
+  isSuccessfulOutboundDocRetrievalResponse,
+  isNonErroringOutboundPatientDiscoveryResponse,
+  isSuccessfulOutboundPatientDiscoveryResponse,
+} from "@metriport/ihe-gateway-sdk";
+
+import { httpErrorCode, schemaErrorCode } from "@metriport/core/external/carequality/error";
+
+export type GatewayCounts = {
+  ehealthexchange: number;
+  epic: number;
+  allscripts: number;
+  ntst: number;
+  redox: number;
+  surescripts: number;
+  athena: number;
+  nextgen: number;
+  kno2: number;
+  medent: number;
+  healthgorilla: number;
+};
+
+export function getDocumentReferenceContentTypeCounts(
+  docRefsContentTypes: string[]
+): Record<string, number> {
+  const contentTypeCounts = docRefsContentTypes.reduce((acc, curr) => {
+    acc[curr] = (acc[curr] || 0) + 1;
+
+    return acc;
+  }, {} as Record<string, number>);
+
+  return contentTypeCounts;
+}
+
+export function determinePatientDiscoverySuccessGateway(
+  response: OutboundPatientDiscoveryResp
+): GatewayCounts {
+  const gateways = [
+    "ehealthexchange",
+    "epic",
+    "allscripts",
+    "ntst",
+    "redox",
+    "surescripts",
+    "athena",
+    "nextgen",
+    "kno2",
+    "medent",
+    "healthgorilla",
+  ];
+
+  const urlLowerCase = response.gateway.url.toLowerCase();
+  const result = gateways.reduce((acc, gateway) => {
+    acc[gateway] = urlLowerCase.includes(gateway);
+    return acc;
+  }, {} as Record<string, boolean>);
+
+  return {
+    ehealthexchange: result.ehealthexchange ? 1 : 0,
+    epic: result.epic ? 1 : 0,
+    allscripts: result.allscripts ? 1 : 0,
+    ntst: result.ntst ? 1 : 0,
+    redox: result.redox ? 1 : 0,
+    surescripts: result.surescripts ? 1 : 0,
+    athena: result.athena ? 1 : 0,
+    nextgen: result.nextgen ? 1 : 0,
+    kno2: result.kno2 ? 1 : 0,
+    medent: result.medent ? 1 : 0,
+    healthgorilla: result.healthgorilla ? 1 : 0,
+  };
+}
+
+export function determinePatientDiscoveryFailureType(response: OutboundPatientDiscoveryResp): {
+  httpError: boolean;
+  schemaError: boolean;
+  specificError: boolean;
+} {
+  if (response.operationOutcome?.issue?.[0]?.code) {
+    const issueCode = response.operationOutcome.issue[0].code;
+    if (issueCode === httpErrorCode) {
+      return { httpError: true, schemaError: false, specificError: false };
+    } else if (issueCode === schemaErrorCode) {
+      return { httpError: false, schemaError: true, specificError: false };
+    }
+  }
+  return {
+    httpError: false,
+    schemaError: false,
+    specificError: true,
+  };
+}
```

Comment:
Why not make these attributes optional, and only return the one that's true in each case? The downstream logic would not even change, but this code will be way cleaner